observability: annotate Session+SessionPool events

odeke-em opened the pull request:
This change adds annotations for session and session pool events to aid customers in debugging latency issues with session pool malevolence and also for maintainers to figure out which session pool type is the most appropriate.

Updates #1170

BEGIN_COMMIT_OVERRIDE
feat: add additional opentelemetry span events for session pool
END_COMMIT_OVERRIDE

odeke-em commented:
## Exhibit
On running examples/grpc_instrumentation_enabled.py we get back these traces
<img width="1554" alt="Screenshot 2024-09-18 at 9 39 29 PM" src="https://github.com/user-attachments/assets/57cc6a24-b8c5-4615-95b7-ec9cf286cecb">
<img width="1542" alt="Screenshot 2024-09-18 at 9 39 47 PM" src="https://github.com/user-attachments/assets/d955ff19-056c-4d88-917d-53e1177bea5c">
<img width="1554" alt="Screenshot 2024-09-18 at 9 40 01 PM" src="https://github.com/user-attachments/assets/cc783f85-56cb-4ae6-9914-091fc851b1a6">
<img width="1555" alt="Screenshot 2024-09-18 at 9 40 25 PM" src="https://github.com/user-attachments/assets/0712976c-56f8-4a80-976d-45a5cffef24d">

odeke-em commented:
cc @harshachinta

odeke-em commented:
Kindly cc-ing you @harshachinta.

odeke-em commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -447,6 +521,12 @@ def get(self, timeout=None):
+        start_time = time.time()
+        current_span = get_current_span()
+        current_span.add_event(
+            "Waiting for a session to become available", {"kind": "pinging_pool"}
```
Because really it is waiting for a session to become available per
https://github.com/googleapis/python-spanner/blob/b2e67626739c05e8dc2eb81e522ff4931f9fac45/google/cloud/spanner_v1/pool.py#L218-L238
and in our document requirements there is a clear need to add that case
<img width="707" alt="Screenshot 2024-10-05 at 10 43 59 PM" src="https://github.com/user-attachments/assets/54c2a310-5612-414d-99ef-02600263be12">

odeke-em commented on google/cloud/spanner_v1/session.py:
```diff
@@ -148,6 +162,7 @@ def create(self):
                 metadata=metadata,
             )
         self._session_id = session_pb.name.split("/")[-1]
+        current_span.add_event("Using Session", {"id": self._session_id})
```
Gotcha and thank you, fixed!

odeke-em commented on google/cloud/spanner_v1/session.py:
```diff
@@ -113,6 +117,10 @@ def name(self):
         :raises ValueError: if session is not yet created
         """
         if self._session_id is None:
+            err = "No session available"
```
You do raise a great point, fixed up and thank you!

odeke-em commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -78,8 +78,17 @@ def trace_call(name, session, extra_attributes=None):
+def set_span_error_and_record_exception(span, exc):
+    if exc and span:
+        span.set_status(Status(StatusCode.ERROR, str(exc)))
+        span.record_exception(exc)
```
Great questions and thank you @harshachinta! So this code is now only used in 1 location which passes in the Exception and no more strings. Good points about the exception being raised within trace_call itself.

odeke-em commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -307,6 +347,10 @@ def get(self):
+
         try:
             session = self._sessions.get_nowait()
         except queue.Empty:
```
Done and thank you!

odeke-em commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -316,6 +360,15 @@ def get(self):
+                        "kind": type(self).__name__,
+                    },
+                )
+
```
Done, thank you @harshachinta!

odeke-em commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -316,6 +360,15 @@ def get(self):
             if not session.exists():
```
I did add the span event down below, thanks!

odeke-em commented:
Kindly please take another look @harshachinta, feedback addressed, thank you for the code review!

odeke-em commented on google/cloud/spanner_v1/session.py:
```diff
@@ -113,6 +117,10 @@ def name(self):
+            err = "No session available"
+            current_span = get_current_span()
+            current_span.add_event(err)
+            set_span_error_and_record_exception(current_span, err)
```
Yeah, I removed the exception recording because as you noted, trace_call already tracks and records exceptions.

odeke-em commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -199,13 +203,32 @@ def bind(self, database):
             )
         self._database_role = self._database_role or self._database.database_role
+        requested_session_count = self.size - self._sessions.qsize()
         request = BatchCreateSessionsRequest(
```
Done, thanks!

harshachinta commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -78,8 +78,17 @@ def trace_call(name, session, extra_attributes=None):
             span.set_status(Status(StatusCode.OK))
+
+
+def set_span_error_and_record_exception(span, exc):
```
Since this is not used anywhere else, we can revert it back as it was.

odeke-em replied:
I did use it in the except block @harshachinta

harshachinta commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -199,21 +203,47 @@ def bind(self, database):
+            )
+
+            if requested_session_count > 0:
+                current_span.add_event(
```
current_span is not initialized.

odeke-em replied:
Sheesh, sorry I had a bunch of benchmarks running locally and they interrupted my `nox -s unit-3.8` tests so didn't see those but all fixed now.

harshachinta commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -199,21 +204,47 @@ def bind(self, database):
+            )
+
+            if requested_session_count > 0:
+                span.add_event(
```
Clearly `span` here could be `None` if opentelemetry is not installed (see https://github.com/googleapis/python-spanner/blob/b2e67626739c05e8dc2eb81e522ff4931f9fac45/google/cloud/spanner_v1/_opentelemetry_tracing.py#L56).

In that case it can throw `AttributeError: 'NoneType' object has no attribute 'add_event'`

odeke-em replied:
@harshachinta I added a guard to it.

harshachinta commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -229,12 +260,24 @@ def get(self, timeout=None):
+            )
             session = self._database.session()
             session.create()
+            span_event_attributes["id"] = self._session_id
```
Here _session_id is not a propery of pool.
Test cases are already failing for this
![image](https://github.com/user-attachments/assets/a1185642-22a6-4f34-a089-77fe7786cf76)

harshachinta replied:
Looks like this should be session.session_id

harshachinta commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -229,12 +260,24 @@ def get(self, timeout=None):
+        span_event_attributes = {"kind": type(self).__name__}
+        current_span.add_event("Acquiring session", span_event_attributes)
         session = self._sessions.get(block=True, timeout=timeout)
+        span_event_attributes["session.id"] = session.session_id
```
Here `session.session_id ` can be empty, if session does not exist. Move this attribute addignment to after the if block below, as it ensures that session will be available.

harshachinta commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -229,12 +260,24 @@ def get(self, timeout=None):
+            )
             session = self._database.session()
             session.create()
+            span_event_attributes["id"] = self._session_id
```
What is the attribute we are using for session id? 
In some place above it was `session.id`, but here it is `id`. Can we follow the attribure naming consistently?

harshachinta commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -229,12 +260,24 @@ def get(self, timeout=None):
+            current_span.add_event("Using Session", span_event_attributes)
 
+        span_event_attributes["time.elapsed"] = time.time() - start_time
+        current_span.add_event("Acquired session", span_event_attributes)
```
It makes more sense to have "Acquired Session" and then "Using Session".

Also I feel "Acquired Session" event should exist when a session lies in pool. If there is no session in pool then we are creating session and in that case "Acquired Session" does not make sense.

So the span hierachy would something be like this.

If session exists in the pool, then
1. Acquiring Session
2. Acquired Session
3. Using Session

If no session in the pool 
1. Acquiring Session
2. No Session Available
3. Creating Session
4. CloudSpanner.CreateSession
5. Using Session

Does that sound good?
@surbhigarg92 Do you have any comments on this structure?

harshachinta commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -422,6 +480,18 @@ def bind(self, database):
             session_template=Session(creator_role=self.database_role),
         )
 
+        requested_session_count = request.session_count
```
Have you not followed the BatchCreateSession span here?

odeke-em replied:
@harshachinta, I did use get_current_span() to retrieve the span down below. This PR is for adding span annotations rather than for creating new spans.

harshachinta commented on google/cloud/spanner_v1/session.py:
```diff
@@ -113,7 +116,10 @@ def name(self):
-            raise ValueError("No session ID set by back-end")
+            err_message = "No session ID set by back-end"
+            current_span = get_current_span()
+            current_span.add_event(err_message)
```
Will this error not get captured by parent span? Do we need to add it as event here?

odeke-em replied:
The error shall be set but not the events on the span aka when it occured. However, now that you mention it, I was using it to debug and should perhaps just remove it entirely as it is non-standard of a span event/annotation.

harshachinta commented on google/cloud/spanner_v1/session.py:
```diff
@@ -124,8 +130,13 @@ def create(self):
         if self._session_id is not None:
-            raise ValueError("Session ID already set by back-end")
+            err_message = "Session ID already set by back-end"
+            current_span.add_event(err_message)
```
Same here also, will this not be captured by `current_span`?

harshachinta commented on tests/unit/test_session.py:
```diff
@@ -174,6 +175,43 @@ def test_create_w_database_role(self):
             "CloudSpanner.CreateSession", attributes=TestSession.BASE_ATTRIBUTES
         )
 
+    def test_create_session_span_annotations(self):
```
Can we add unit tests to verify spans for different pool class similar to what is done for create_session?

odeke-em replied:
Please take a look at the additions to tests/unit/test_pool.py @harshachinta

odeke-em commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -199,21 +203,47 @@ def bind(self, database):
-            resp = api.batch_create_sessions(
-                request=request,
-                metadata=metadata,
+        with trace_call("CloudSpanner.BatchCreateSessions", self) as span:
```
It is imported @harshachinta on line 30.

odeke-em commented:
@harshachinta @surbhigarg92, howdy howdy? Kindly please help me run the bots for this PR. Thank you!

odeke-em commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -78,8 +78,17 @@ def trace_call(name, session, extra_attributes=None):
+
+
+def get_current_span():
+    return trace.get_current_span()
```
anywhere it is used, I've added None checks. Once we upgrade to only support Python3.8 we can then remove it.

odeke-em commented:
@harshachinta @sakthivelmanii this PR is ready for review, kindly please take a look and thank you!

odeke-em commented:
@harshachinta kindly help run the bots for this.

odeke-em commented:
@harshachinta @surbhigarg92 @sakthivelmanii kindly help me trigger the bots for this PR too. Thank you.

odeke-em commented:
This PR could use some bots runs. Thank you!

odeke-em commented:
This PR could use some bots runs. Thank you!

odeke-em committed:
observability: annotate Session+SessionPool events

This change adds annotations for session and session pool
events to aid customers in debugging latency issues with
session pool malevolence and also for maintainers to
figure out which session pool type is the most appropriate.

Updates #1170

harshachinta commented on google/cloud/spanner_v1/database.py:
```diff
@@ -726,6 +730,15 @@ def execute_pdml():
+            with SessionCheckout(self._pool) as session:
+                observability_options = getattr(self, "observability_options", None)
+                with trace_call(
+                    "CloudSpanner.execute_pdml",
```
```suggestion
                    "CloudSpanner.execute_partitioned_dml",
```

harshachinta commented on google/cloud/spanner_v1/database.py:
```diff
@@ -694,7 +698,7 @@ def execute_partitioned_dml(
         def execute_pdml():
-            with SessionCheckout(self._pool) as session:
+            def do_execute_pdml(session):
                 txn = api.begin_transaction(
```
wrap this in trace_call to create a new span `CloudSpannerOperation.BeginTransaction`?
Ideally, the traces for pdml will look like,
+ CloudSpanner.execute_partitioned_dml
      + Some session related span if there are any
      +. CloudSpannerOperation.BeginTransaction
      + CloudSpannerOperation.ExecuteStreamingSql

harshachinta commented on google/cloud/spanner_v1/database.py:
```diff
@@ -726,6 +730,15 @@ def execute_pdml():
 
                 return result_set.stats.row_count_lower_bound
 
+            with SessionCheckout(self._pool) as session:
```
I am not able to add comment in above line, hence adding comment here
pass `trace_name="CloudSpannerOperation.ExecuteStreamingSql"` at https://github.com/googleapis/python-spanner/blob/054a18658eedc5d4dbecb7508baa3f3d67f5b815/google/cloud/spanner_v1/database.py#L717

odeke-em replied:
The code from PR #1241 crept into this PR, I am going to revert it.

harshachinta commented on google/cloud/spanner_v1/database.py:
```diff
@@ -880,17 +893,26 @@ def run_in_transaction(self, func, *args, **kw):
+
+                # Check out a session and run the function in a transaction; once
+                # done, flip the sanity check bit back.
+                try:
```
The try block should start before with SessionCheckout since that can also throw error

harshachinta commented on google/cloud/spanner_v1/database.py:
```diff
@@ -726,6 +730,15 @@ def execute_pdml():
 
                 return result_set.stats.row_count_lower_bound
 
+            with SessionCheckout(self._pool) as session:
```
why is the trace_call started after SessionCheckout?
Since session checkout is part of this method, we should start the span before this with block itself.
Then only session related operations such as getting session etc will be part of pdml parent span.

harshachinta commented on google/cloud/spanner_v1/database.py:
```diff
@@ -880,17 +893,26 @@ def run_in_transaction(self, func, *args, **kw):
-            self._local.transaction_running = False
+        with SessionCheckout(self._pool) as session:
+            observability_options = getattr(self, "observability_options", None)
+            with trace_call(
```
similar to above, trace_call should start before SessionCheckout

harshachinta commented on google/cloud/spanner_v1/snapshot.py:
```diff
@@ -538,7 +538,7 @@ def _get_streamed_result_set(
             restart,
             request,
-            "CloudSpanner.ReadWriteTransaction",
+            "CloudSpanner.execute_sql",
```
```suggestion
            "CloudSpanner.execute_streaming_sql",
```
since it is a streaming call

odeke-em commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -77,10 +77,11 @@ def trace_call(name, session, extra_attributes=None, observability_options=None)
         "db.type": "spanner",
         "db.url": SpannerClient.DEFAULT_ENDPOINT,
-        "db.instance": session._database.name,
+        "db.instance": "" if not db else db.name,
```
During testing, this scenario happens when unit testing trace_call

odeke-em replied:
We also found a purpose for it in PR #1241 where session might be None, though in that PR we handle deriving it from observability_options too.

odeke-em commented on google/cloud/spanner_v1/database.py:
```diff
@@ -1187,6 +1193,11 @@ def __exit__(self, exc_type, exc_val, exc_tb):
                     extra={"commit_stats": self._batch.commit_stats},
                 )
             self._database._pool.put(self._session)
+            current_span = get_current_span()
+            if current_span:
+                current_span.add_event(
+                    "Returned session to pool", {"id": self._session.session_id}
+                )
```
This gets quickly expensive if we use it in more than 1 place.

odeke-em replied:
I did however introduce a helper that used the current span and added events to remove the `if truth` check

odeke-em commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -106,3 +107,7 @@ def trace_call(name, session, extra_attributes=None, observability_options=None)
+
+
+def get_current_span():
+    return trace.get_current_span()
```
Done and thanks!

odeke-em commented:
Kindly take a look again @harshachinta and help me run the bots too @sakthivelmanii. Thank you!

harshachinta commented on tests/unit/test_pool.py:
```diff
@@ -198,6 +213,63 @@ def test_get_non_expired(self):
+        wantEventNames = [
+            "Requesting 1 sessions",
+            "Creating 1 sessions",
+            "exception",
+            "exception",
```
why are there 2 exception events?

odeke-em replied:
It is a quirk with opentelemetry-python; they add "exception" on every __exit__ call whether or not you set your own, I'll file a bug later on but for now I'll just remove our explicit invocation of "span.record_exception" and instead add our comment.

harshachinta commented on tests/unit/test_pool.py:
```diff
@@ -198,6 +213,63 @@ def test_get_non_expired(self):
             self.assertTrue(session._exists_checked)
             self.assertFalse(pool._sessions.full())
 
+    def test_spans_bind_get(self):
```
can we add one more negative case for fixedsizepool get method?
scenario - Session pool is empty, and there is a get call on the pool. In this case it will wait until timeout and then fail

odeke-em replied:
Done!

harshachinta commented on tests/unit/test_pool.py:
```diff
@@ -328,6 +408,31 @@ def test_get_empty(self):
         session.create.assert_called()
         self.assertTrue(pool._sessions.empty())
 
+    def test_spans_get_empty_pool(self):
```
can you add one line comment for each test on the scenario it verifies?

harshachinta commented on tests/unit/test_pool.py:
```diff
@@ -342,6 +447,24 @@ def test_get_non_empty_session_exists(self):
+            self.assertTrue(session._exists_checked)
+            self.assertTrue(pool._sessions.empty())
+
+        self.assertSpanAttributes(
```
can you also add assertion on span events here?

odeke-em replied:
Done, thanks!

harshachinta commented on tests/unit/test_pool.py:
```diff
@@ -655,9 +832,45 @@ def test_ping_oldest_stale_and_not_exists(self):
+        self.assertSpanEvents("pool.Get", wantEventNames)
+
+
+class TestTransactionPingingPool(OpenTelemetryBase):
```
you can remove tests from TransactionPingingPool since it is deprecated

odeke-em replied:
Done, thanks!

odeke-em commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -421,7 +499,31 @@ def bind(self, database):
+            span_event_attributes,
+        )
+
+        if created_session_count >= self.size:
```
I had it in my tests as a condition for a reused pool but sure I can delete it.

odeke-em commented on tests/_helpers.py:
```diff
@@ -16,10 +16,11 @@
     from opentelemetry.trace.status import StatusCode
 
-    trace.set_tracer_provider(TracerProvider())
+    trace.set_tracer_provider(TracerProvider(sampler=TraceIdRatioBased(1.0)))
```
Yes! Making it explicit to be a 100% sampler removes all flakes; I went through headaches during tests for it; better to be explicit.

odeke-em commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -229,12 +263,37 @@ def get(self, timeout=None):
+            )
             session = self._database.session()
             session.create()
+            span_event_attributes["session.id"] = session._session_id
```
Well firstly, in line 283 we had the original session_id that was invalid; it isn't a mistake to repalce the new session.id given that now we've recreated it. The new suggestion would not show it when reporting "Session is not valid, recreating it".

odeke-em commented:
Thank you @harshachinta for the review, I've addressed the feedback. Please take a look again.

harshachinta reviewed (approved):
Please fix lint

odeke-em commented on google/cloud/spanner_v1/_opentelemetry_tracing.py:
```diff
@@ -102,7 +103,21 @@ def trace_call(name, session, extra_attributes=None, observability_options=None)
             yield span
         except Exception as error:
             span.set_status(Status(StatusCode.ERROR, str(error)))
-            span.record_exception(error)
```
It's because OpenTelemetry was upgraded only recently.

odeke-em replied:
We do have locked tests to check for the exceptions to ensure that they are in there and from Span.__enter__. I had to dive back into OpenTelemetry-Python's code as it isn't even documented and in our demos it was very distracting to have mysteriously both errors. I think for the sake of our sanity and project stability let's leave that comment in and if anything happens it is a trivial one to add back @harshachinta

harshachinta replied:
Hmm. But the opentelemetry documentation for Python guides to record exception for instrumentation libraries.
https://opentelemetry.io/docs/languages/python/instrumentation/#record-exceptions-in-spans

Can you share the code pointer on where the opentelemetry records exception by default when exiting?

odeke-em replied:
@harshachinta https://github.com/open-telemetry/opentelemetry-python/blob/db4ef06f471cb9e94d810ae66964c3e5e0f5e61e/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L986-L1008

odeke-em replied:
@harshachinta it was the cause of us seeing 2 exceptions and took a ton of confusion and time for me to debug, they don't seem to document this condition.

harshachinta commented on tests/unit/test_pool.py:
```diff
@@ -655,193 +874,36 @@ def test_ping_oldest_stale_and_not_exists(self):
+        self.assertNoSpans()
 
-
-class TestTransactionPingingPool(unittest.TestCase):
```
I didn't mean to remove the existing tests - TestTransactionPingingPool. Can you revert this back?
Since this was already deprecated we need not any observability related tests here.

odeke-em replied:
Okay gotcha. Sure let me make the update shortly.

odeke-em commented:
@harshachinta kindly help me re-run the bots; all unit tests pass locally.
